### PR TITLE
Improve utils.get nodes

### PR DESCRIPTION
**Pull request by fbudin69500**

### CODE REVIEW

**Comment by jcfr** [comment]
Thanks for the patch :+1: 

Cc: @Slicer/slicer-core @Slicer/community

**Comment by hjmjohnson** [comment]
👍

File: Base/Python/slicer/tests/test_slicer_util_getNodes.py [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
     def setUp(self):
         slicer.mrmlScene.AddNode(slicer.vtkMRMLScalarVolumeNode()).SetName("Volume1")
         slicer.mrmlScene.AddNode(slicer.vtkMRMLScalarVolumeNode()).SetName("Volume2")
+        slicer.mrmlScene.AddNode(slicer.vtkMRMLScalarVolumeNode()).SetName("Volume")
+        slicer.mrmlScene.AddNode(slicer.vtkMRMLScalarVolumeNode()).SetName("Volume")
+
+    def test_getNodesMultipleNodesSharingName(self):
+
+        self.assertTrue("Volume" in slicer.util.getNodes("Volume"))
+        self.assertTrue("Volume" in slicer.util.getNodes("Volume",useLists=True))
```
Comment (msmolens) [inline Base/Python/slicer/tests/test_slicer_util_getNodes.py]: These can be written as [`self.assertIn(...)`](https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.assertIn) to be easier to read and to provide better error messages on failure.

File: Base/Python/slicer/tests/test_slicer_util_getNodes.py [hunk 2]
```diff
@@ -1,11 +1,24 @@
...
+
+        self.assertEqual(slicer.util.getNodes("Volume").keys(), ["Volume"])
+        self.assertEqual(isinstance(slicer.util.getNodes("Volume")["Volume"], vtk.vtkObject), True)
```
Comment (msmolens) [inline Base/Python/slicer/tests/test_slicer_util_getNodes.py]: This line, and similar lines, can be written as [`self.assertIsInstance(...)`](https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.assertIsInstance).

File: Base/Python/slicer/tests/test_slicer_util_getNodes.py [hunk 3]
```diff
@@ -17,4 +30,7 @@ def test_getNodes(self):
 
         self.assertEqual(slicer.util.getNodes("Volume1").keys(), ["Volume1"])
         self.assertEqual(slicer.util.getNodes("Volume2").keys(), ["Volume2"])
-        self.assertEqual(slicer.util.getNodes("Volume*").keys(), ["Volume1", "Volume2"])
+        self.assertEqual(slicer.util.getNodes("Volume*").keys(), ["Volume", "Volume1", "Volume2"])
+
+    def test_getFirstNodeByName(self):
+        self.assertEqual(slicer.util.getFirstNodeByName("Volume",'vtkMRMLScalarVolumeNode').GetName(),"Volume1" )
```
Comment (msmolens) [inline Base/Python/slicer/tests/test_slicer_util_getNodes.py]: Can you add a space following the commas, here and elsewhere?

File: Base/Python/slicer/util.py [hunk 4]
```diff
@@ -458,9 +458,12 @@ def resetSliceViews():
 # MRML
 #
 
-def getNodes(pattern = "", scene=None):
+def getNodes(pattern = "", scene=None, useLists=False):
     """Return a dictionary of nodes where the name or id matches the 'pattern'.
-    Providing an empty 'pattern' string will return all nodes.
+    Providing an empty 'pattern' string will return all nodes. If multiple
+    node share the same name, using 'useLists=False' (default behavior) will
```
Comment (msmolens) [inline Base/Python/slicer/util.py]: > If multiple node

Should be plural

File: Base/Python/slicer/util.py [hunk 5]
```diff
@@ -474,7 +477,11 @@ def getNodes(pattern = "", scene=None):
       if (pattern == "" or
           fnmatch.fnmatchcase(name, pattern) or
           fnmatch.fnmatchcase(id, pattern)):
-        nodes[node.GetName()] = node
+        if useLists:
+          print("getNodes name: %s"%name)
```
Comment (msmolens) [inline Base/Python/slicer/util.py]: Leftover debugging output?

File: Base/Python/slicer/util.py [hunk 6]
```diff
@@ -503,14 +510,9 @@ def getFirstNodeByName(name, className=None):
...
-  for nodeName in nodes.keys():
-    if not className:
-      return (nodes[nodeName]) # return the first one
-    else:
-      if nodes[nodeName].IsA(className):
-        return (nodes[nodeName])
-  return None
+  import slicer
+  scene = slicer.mrmlScene
+  return scene.GetFirstNode(name,className,[0],False)
```
Comment (msmolens) [inline Base/Python/slicer/util.py]: Add spaces after the commas.

File: Base/Python/slicer/util.py (hunk 5 above)
Comment (lassoan) [inline Base/Python/slicer/util.py]: For logging use logging.debug (only written to log, not to the console), logging.info, etc. See https://www.slicer.org/slicerWiki/index.php/Documentation/Nightly/Developers/Style_Guide#Logging for details.

**Comment by fbudin69500** [comment]
Thanks for the reviews!

**Comment by jcfr** [comment]
Re-organized a little, and also fixed an issue in `getNode`.

Integrated in [r25153](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=25153), [r25154](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=25154), [r25155](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=25155), [r25156](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=25156)

**Pull request closed without merging**